openstax/rhaptos.cnxmlutils
# keep footnotes in content (partial fix for OpenStax books)

Before, footnotes were collated into a Footnotes section at the bottom of every page. This keeps the footnotes in the content but it has the downside that `<c:footnote>` elements allowed blockish elements in them (e.g. `<c:para>` or `<c:list>`.

Partial fix for https://github.com/Connexions/cnx-recipes/issues/566

# TODO

- [x] unwrap `<c:para>` elements (figure out what to do about whitespace)
- [x] figure out what to do about `<c:list>` elements. Should we inline them? ~~or does XHTML allow `<ol>` inside a `<span>`?~~
- [x] look at being more standards-compliant (will need for GDocs import)

## Standards-compliant notes

Result should convert something like this:

```xml

<c:para>In that year<c:footnote id="ft2f"><c:para>
     2 The manufacturing statistics for 1900 which
     follow are not those given in the Twelfth
     Census, but are taken from the 
     <c:emphasis>Census of Manufactures</c:emphasis> …
   </c:para></c:footnote>
   there were 67 mills engaged in the manufacture of
   cotton goods …</c:para>
```

into something like this:

```html
<p>
   In that
   year<a href="#ft2f" role="doc-noteref" epub:type="noteref">[footnote]</a>
   there were 67 mills engaged in the manufacture of
   cotton goods …
</p>
<aside id="ft2f" role="doc-footnote" epub:type="footnote">
   <p>
     2 The manufacturing statistics for 1900 which
     follow are not those given in the Twelfth
     Census, but are taken from the 
     <em>Census of Manufactures</em> …
   </p>
</aside>
```

- [DAISY/EPUB guidelines for footnotes that are unlinked from the content](http://kb.daisy.org/publishing/docs/html/notes.html)
- [Apple's iBooks on footnotes](https://help.apple.com/itc/booksassetguide/en.lproj/itccf8ecf5c8.html)
- [Pandoc's HTML Reader code](https://github.com/jgm/pandoc/blob/652ed0b82cd7095f418859356d7e5f8ada65eb49/src/Text/Pandoc/Readers/HTML.hs#L224)
- [GDoc spike notes](https://github.com/openstax/gdoc-spike/blob/master/pandoc-results.md#footnotes)
- Abandoned Pandoc PR/Issue with link @ bottom to current Issue: `https://github.com/jgm/pandoc/pull/4491`


Also, when choosing an element, use `<aside>` (Kindle & iBooks, and pandoc seems to use any element now for importing): `https://github.com/jgm/pandoc/issues/1995#issuecomment-375925048`

# Validation Code

Steps to reproduce:

1. Rename ` rhaptos/cnxmlutils/xsl/test/footnote.cnxml.html` to be `.xhtml`
1. add the following to the top of the file
1. open in the Firefox browser
1. check the dev console for errors

<details><summary>Click me to see the code</summary>

```html
<html
  xmlns="http://www.w3.org/1999/xhtml"
  xmlns:epub='http://www.idpf.org/2007/ops'>
<head>
  <title>Validate ids and links</title>
  <style>:target {background-color: #ff0;}</style>
  <script>document.addEventListener('DOMContentLoaded', () => {
    // ensure ids are unique
    const ids = new Set()
    Array.from(document.querySelectorAll('[id]')).forEach(el => {
      const id = el.getAttribute('id')
      if (ids.has(id)) { throw new Error(`Duplicate id: ${id}`)}
      ids.add(id)
    })
    console.log(`unique id attributes: ${ids.size}`)
  
    // ensure all links point to something
    const links = new Set()
    Array.from(document.querySelectorAll('[href^="#"]')).forEach(el => {
      const id = el.getAttribute('href').slice(1)
      if (!ids.has(id)) { throw new Error(`Link points to nowhere: ${id}`)}
      links.add(id)
    })
    console.log(`internal links: ${links.size}`)
  
    // ensure all id="linktome..." have a link pointing to them
    let linktomecount = 0
    Array.from(ids).forEach(id => {
      if (id.startsWith('linktome')) {
        linktomecount++
        if (!links.has(id)) { throw new Error(`Nothing linked to ${id}`)}
      }
    })
    if (linktomecount === 0) { throw new Error('there were no elements with an id beginning with "linktome-"')}
    console.log(`linktome elements: ${linktomecount}`)
    alert('Tests ran and seem to have passed')
  
  })</script>
</head>
```

</details>
